### DISCUSSION TITLE
DSP and generating better levels

### DISCUSSION THREAD

**Discussion by pserwylo** [beat-feet/beat-feet#43] [Q&A]
Currently, the game world is generated by doing more or less the following:
* Reading PCM samples from an MP3 file (currently all 44100Hz)
* Passing them through an FFT with a window of size 1024
* Converting the complex output of the FFT into intensity values by taking the absolute value of them
* Smoothing the values using a moving-median (seemed to produce slightly nicer curves than a moving-average)
* Running a (very poor excuse of) differentiation over the resulting curve (just look for points in the graph where the median is either increasing or decreasing, and measure the time in between as a local maximum.
* Do some processing of these blocks of time and strengths, so that some of them which are close in size are merged together, and if they are bigger than a single tile, they are rounded up to the nearest tile.

Once we have these sizes, they are converted into buildings, fences, etc based on their size.

However, it still doesn't quite feel just right. I am hoping to get it so that every jump really feels like it is timed with the beat of the music.

Below is a video of the little playground app I created to analyse the music (can be run by executing the desktop libgdx app with the "playground" cli argument). It shows a spectogram, along with a bunch of various statistics pulled from the spectogram (e.g. average frequency, stddev of frequency over a window, etc). The final column in this is the poor-effort at differentiating two of the other series to identify something like features of interest.

https://user-images.githubusercontent.com/248565/104392871-2db21400-5597-11eb-8387-7c9db9ae9a62.mp4

**Comment by pserwylo** [comment]
Since writing, this has been changed slightly to perform the same analysis above, but to do it on three different spectograms. One for the lowest third of frequencies in the FFT, one for the mid range frequencies, and one for the high frequencies. All obstacles are combined together and treated the same, but it may be nice to do something different with high frequency features compared to low frequency, etc.

In practice, it seems to make for more interesting levels.

**Reply by bydariogamer** [reply]
I made a ["beat-game" PC version](https://github.com/bydariogamer/beatgame) using `pygame` for a game jam... The game doesn't resemble to this one but in the idea of procedural level generation... I name it as your game as a tribute :wink:
High level programming languages are too wasteful to make a serious amount of transformations in an acceptable period of time (I tried to load big music file in beat-feet and I had to stop it after five minutes... Do you know the O of your algorithm?)
So, I tried to make an algorithm just based on the average intensity of the samples per half-second, but it was a bit _impossible_. The levels are boring... Your algorithm is definitely way better than any simple one that uses trivial low time (I think it is O(n^2) but the truth is a 5 minutes song took just (average) 1.46 seconds to load).

**Reply by Sogolumbo** [reply]
I just implemented a beat finder in my fork of your game. It needs a little more fine adjustment, then other features can be built on that.

**Reply by pserwylo** [reply]
Thanks to both of you. I'll try to check this out soon. I'm a big believe that anything can be written in any language, so if it turns out this is the way forward, I'm happy to put the time in to migrate it to Java if need be.

Oh, and regarding the big-O of my algorithm, no, I have no idea. I just kept hacking away until empirically it seemed to work well :(

**Comment by Sogolumbo** [comment]
Maybe it's possible to find the tempo (bpm) of the song by using autocorrelation.
Then you have the basic "grid" of the music and their subdivisions (e.g. sixteenth notes).
Based on that you could evaluate the your metric (median loudness or whatever) per subdivision so that all obstacles are right on time.

By changing the amount of subdivisions (and applying rules to the map) you can also tweak the difficulty of the song.

This will only (easily) work if the songs have a constant tempo. (It might be possible to do this with a variable tempo but it's more resource intensive and more prone to error).

**Reply by Sogolumbo** [reply]
One more idea: When you have the tempo you could check how much difference there is between the current bar and the previous bar. Then you could amplify that difference. This would (hopefully) result in the "interesting part" (changing melodies, changing beats) playing a bigger role.

**Comment by JartThrower** [comment]
I suggest you abandon automatic generation of maps in favor of a manual process.

If you look at other rhythm games you'll see that most of them use manually created maps for their included songs. The reason for that is that making an algorithm that will produce good, playable maps that match the beat of a song is a very hard problem. In the end it will take more effort than simply mapping out each song manually and likely produce worse results.
